core/common/src/main/java/alluxio/AbstractThriftClient.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio;
+
+import alluxio.exception.AlluxioException;
+import alluxio.thrift.AlluxioService;
+import alluxio.thrift.AlluxioTException;
+import alluxio.thrift.ThriftIOException;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * The base class for clients that use {@link alluxio.network.connection.ThriftClientPool}.
+ *
+ * @param <C> the Alluxio service type
+ */
+@ThreadSafe
+public abstract class AbstractThriftClient<C extends AlluxioService.Client> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final int RPC_MAX_NUM_RETRY = 30;
+
+  /**
+   * @return a Thrift service client
+   */
+  protected abstract C acquireClient() throws IOException;
```

Comment:
yes, good point. that is better to be documented. if acquireClient always returns the same ThriftClient then it is not threadsafe